Refactor observables doc for style, clarity, and structure (#38165)
angular/angular: aio/content/guide/observables.md

RxJS provides an implementation of the `Observable` type, which makes it easier to compose asynchronous or callback-based code in TypeScript applications. The library also provides many useful [operators and utility functions](#rx-library "Using RxJS functions and operators") for working with observables.

## How observables work

Observables get their name from the observer _pattern_. In this pattern, an object maintains a list of its dependents, called *observers*, and notifies them automatically of state changes.
This pattern is similar to the [publish/subscribe](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern "Wikipedia") design pattern.

Observables are declarative&mdash;that is, you define a function for publishing values, but it is not executed until a consumer subscribes to it.
The subscribed consumer then receives notifications until the function completes or errors out, or until they unsubscribe.

An observable can deliver multiple values of any type&mdash;literals, messages, or events, depending on the context.
Consumers can receive observable values synchronously or asynchronously.

> You subscribe to an observable to consume values, and unsubscribe when the execution completes or you no longer need to listen for new values.

Reviewer: I think it's worth expanding on the concepts here, so I've taken a stab at this:

You subscribe to an observable to begin receiving values from it, and can unsubscribe when you no longer need to listen for new values. Some observables (such as those from `HttpClient`) are finite, and will unsubscribe automatically upon completion.

# Using observables to pass values

The Angular API makes frequent use of observables, and applications commonly use them for event handling and asynchronous programming. Some parts of the Angular API expect or return values in observables, and you can use them in your own code to pass messages between parts of your application.

Angular makes use of and extends Reactive Extensions for JavaScript (RxJS), a library for [reactive programming](https://en.wikipedia.org/wiki/Reactive_programming "Wikipedia").
RxJS provides an implementation of the `Observable` type, which makes it easier to compose asynchronous or callback-based code in TypeScript applications. The library also provides many useful [operators and utility functions](#rx-library "Using RxJS functions and operators") for working with observables.

## How observables work

> Observables get their name from the observer _pattern_. In this pattern, an object (sometimes called the *subject*) maintains a list of its dependents, called *observers*, and notifies them automatically of state changes.

Reviewer 2: I honestly think this might confuse many readers or at least moves them in the wrong direction as Subjects and Observables are fundamentally different in RxJS.

This pattern is similar to the [publish/subscribe](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern "Wikipedia") design pattern.

Observables are declarative&mdash;that is, you define a function for publishing values, but it is not executed until a consumer subscribes to it.
The subscribed consumer then receives notifications until the function completes or errors out, or until they unsubscribe.

An observable can deliver multiple values of any type&mdash;literals, messages, or events, depending on the context.
Consumers can receive observable values synchronously or asynchronously.
You subscribe to an observable to consume values, and unsubscribe when the execution completes or you no longer need to listen for new values.

> The interface for listening to values and stopping listening is the same, regardless of whether the value stream reports keystrokes, an HTTP response, or the state of an interval timer.

Reviewer 2: "The interface for listening to values and stopping listening is the same" I don't understand this?

An observer object can define any combination of these handlers. If you don't supply a handler for a notification type, the observer ignores notifications of that type.

## Subscribing to observables

> An `Observable` instance begins publishing values only when you subscribe to it.

Reviewer: It may be worth including here that subscribing may cause the observable to take some action to begin producing values - for example, subscribing to an `HttpClient` observable causes an HTTP request to be made.

You subscribe by calling the `subscribe()` method of the observable instance, passing an observer object to receive the notifications.

The following example creates and subscribes to a basic observable, with an observer that logs the received message to the console.

<div class="alert is-helpful">

For demonstration purposes, this example creates the observable using an [RxJS creation function](#observable-creation-functions "Read more about creation functions").
The `of(...items)` function returns an Observable instance that synchronously delivers the values provided as arguments.

You can also use a constructor to create new instances of `Observable`, as shown in the section [Creating observables](#creating-observables).
</div>

<code-example
  path="observables/src/subscribing.ts"
  region="observer"
  header="Subscribe using observer"></code-example>

> As an alternative to passing in an observer object, the `subscribe()` method can accept inline callback function definitions for `next`, `error`, and `complete` handlers, in that order. For example, the following `subscribe()` call is the same as the one that specifies the predefined observer:

Reviewer: Is it worth switching this example around? In my experience subscribing using an observer is fairly uncommon, and the majority of applications will use the inline form.

The following example creates and subscribes to a basic observable, with an observer that logs the received message to the console.

<div class="alert is-helpful">

For demonstration purposes, this example creates the observable using an [RxJS creation function](#observable-creation-functions "Read more about creation functions").
The `of(...items)` function returns an Observable instance that synchronously delivers the values provided as arguments.

You can also use a constructor to create new instances of `Observable`, as shown in the section [Creating observables](#creating-observables).
</div>

<code-example
  path="observables/src/subscribing.ts"
  region="observer"
  header="Subscribe using observer"></code-example>

As an alternative to passing in an observer object, the `subscribe()` method can accept inline callback function definitions for `next`, `error`, and `complete` handlers, in that order. For example, the following `subscribe()` call is the same as the one that specifies the predefined observer:

<code-example path="observables/src/subscribing.ts" region="sub_fn" header="Subscribe with positional arguments"></code-example>

> In either case, a `next` handler is required. The `error` and `complete` handlers are optional.

Reviewer 2: Not part of this pr, but this sentence is actually wrong. One can also just call `subscribe()` without any callback

Author: What happens to the published values if there is no next handler? Should this option be mentioned?

Reviewer 2: so if you have no handler registered it will not be handled, but produced. So in angular, especially with the `async`pipe I often see people doing any kind of side effect in `tap` and actually not subscribing manually anymore therefore I would like this to be corrected. I am not sure what level of information should be part of this document, but probably @aikidave has an opinion on this

Author: This does seem a bit out of scope, especially this early in the description. I took out the optional/required language.

<code-example path="observables/src/creating.ts" region="subscriber" header="Create observable with constructor"></code-example>

You can create an observable that publishes events. In the following example, the subscriber function is defined inline.

<code-example path="observables/src/creating.ts" region="fromevent" header="Create with custom fromEvent function"></code-example>

You can use this function to create an observable that publishes keydown events, as follows:

<code-example path="observables/src/creating.ts" region="fromevent_use" header="Use custom fromEvent function"></code-example>

> ## Multicasting  values to multiple subscribers

Reviewer: ```suggestion
## Multicasting values to multiple subscribers
```

*Multicasting* is the practice of broadcasting emitted values from a single observable execution to a list of multiple subscribers.

A typical observable creates a new, independent execution for each subscribed observer.
When an observer subscribes, the observable creates and runs a producer instance that delivers values to that observer.
When a second observer subscribes, the observable creates and runs a new producer instance that delivers values to that second observer in a separate execution.

Sometimes, instead of starting an independent execution for each subscriber, you want each subscription to get the same values&mdash;even if values have already started emitting.
For example, you might want to create an observable of clicks on a document object.
With a multicasting observable, you don't register multiple listeners on the document, but instead re-use the first listener and send values out to each subscriber.

> Consider the following example that counts from 1 to 3, with a one-second delay after each number emitted.

Reviewer: It might be valuable to use `HttpClient` observables instead, since this offers a real-world example of when multicasting might be valueable.

A single HttpClient observable, subscribed twice -> two separate HTTP requests
A single HttpClient observable, multicasted and subscribed twice -> a single HTTP request with two observers

Author: I added the description of the more realistic case, but left in the old example that shows how to make the observable multicasting.  

Can you make the observable returned by an `HttpClient` call multicasting? 
Again, the example came from someone else, and I can't create an equivalent -- do you want to try?

<code-example path="observables/src/multicasting.ts" region="delay_sequence" header="Create a delayed sequence"></code-example>

> When creating the observable, you should determine how you want to use that observable and whether you want to multicast its values.

Reviewer: It's not actually when creating the observable, but when subscribing. You might want to multicast observables that you didn't create yourself (e.g. those from `HttpClient`).

If you subscribe twice, notice that there are two separate streams, each emitting values every second.

<code-example path="observables/src/multicasting.ts" region="subscribe_twice" header="Two subscriptions"></code-example>

> The following code changes the observable to use multicasting instead:

Reviewer: Looking through the examples here, I think unfortunately they all really miss the point of multicasting.

The example actually has nothing to do with multicasting. It defines a custom "hot" observable which treats subsequent subscribers differently than the first one. This is not how most applications would make use of multicasting.

Most applications will deal with multicasting by using operators to share a single (unicast?) observable subscription among multiple subscribers, typically with the `multicast` or `share` operators (with which one can choose a specific `Subject` implementation to fine tune the behavior).

So a better set of examples would be:

1. One showing the problem of duplicate requests when subscribing to an HTTP request observable in two places
2. Using the `share()` operator to turn the observable into a multicast observable, ensuring that only one request is made for both subscriptions
3. Using the `shareReplay()` operator to ensure that late subscribers still receive the response, even after the HTTP request is complete.

Author: Can you or someone please supply these examples?

<code-example path="observables/src/creating.ts" region="subscriber" header="Create observable with constructor"></code-example>

You can create an observable that publishes events. In the following example, the subscriber function is defined inline.

<code-example path="observables/src/creating.ts" region="fromevent" header="Create with custom fromEvent function"></code-example>

You can use this function to create an observable that publishes keydown events, as follows:

<code-example path="observables/src/creating.ts" region="fromevent_use" header="Use custom fromEvent function"></code-example>

## Multicasting  values to multiple subscribers

*Multicasting* is the practice of broadcasting emitted values from a single observable execution to a list of multiple subscribers.

A typical observable creates a new, independent execution for each subscribed observer.

> When an observer subscribes, the observable wires up an event handler and delivers values to that observer.

Reviewer 2: Event handler might be a little bit irritating. I think what you are referring to is a Producer instance?

For example, you might want to create an observable of clicks on a document object.
With a multicasting observable, you don't register multiple listeners on the document, but instead re-use the first listener and send values out to each subscriber.

Consider the following example that counts from 1 to 3, with a one-second delay after each number emitted.

<code-example path="observables/src/multicasting.ts" region="delay_sequence" header="Create a delayed sequence"></code-example>

When creating the observable, you should determine how you want to use that observable and whether you want to multicast its values.
If you subscribe twice, notice that there are two separate streams, each emitting values every second.

<code-example path="observables/src/multicasting.ts" region="subscribe_twice" header="Two subscriptions"></code-example>

The following code changes the observable to use multicasting instead:

<code-example path="observables/src/multicasting.ts" region="multicast_sequence" header="Create a multicast subscriber"></code-example>

### Multicast with the _Subject_ class

> RxJS provides the [`Subject` class](https://rxjs.dev/guide/subject "RxJS API reference") that simplifies the process of multicasting, allowing you to take any observable and make it multicasting.

Reviewer 2: Maybe it's just the fact that I am not a native speaker but "allowing you to take any observable and make it multicasting" kinda sounds like an Observable is a constructor argument for a Subject.

<code-example path="observables/src/multicasting-subject.ts" region="multicast_subject" header="Use a Subject for multicasting"></code-example>

## Error handling

> An observable can either produce values (calling the `next` callback), or it can complete, calling either the `complete` or `error` callback.

Reviewer 2: technically an Observable can produce values and complete afterwards or error and it can also do none of those :D

Author: Is this true:  "It can also do none of these things, producing only side effects." 

 Is this why the "next" handler is not required? How would this work? Example?

Should there be a section discussing this usage? (Presumably it wouldn't go in the "Error handling" section.)

Reviewer 2: yes it is true, but the next handler is not required just from an API design perspective, maybe you are not interested in the next values but just error or complete or you really just want to perform the side effects and therefore you just want to subscribe to an observable. 

I am not really sure if this should be covered in more detail as this is very specific and not really help with the overall understanding of rxjs. I just wanted to clarify this information here.

> Because observables produce values asynchronously, the try/catch construction does not effectively catch errors. Instead, you generally handle errors by specifying an `error` callback on the observer.

Reviewer 2: Values are not produced asynchronously, at least not if it isn't necessary. But errors are always emitted asynchronously

You can import these from the `rxjs` library in your Angular application, and use them to accomplish the following kinds of tasks:

* Convert existing code for async operations into observables
* Iterate through the values in a stream
* Map values to different types
* Filter streams
* Compose multiple streams

### Observable creation functions

RxJS offers a number of functions that can be used to create new observables. These functions can simplify the process of creating observables from things such as events, timers, promises, and so on. For example:

<code-example path="rx-library/src/simple-creation.ts" region="promise" header="Create an observable from a promise"></code-example>

<code-example path="rx-library/src/simple-creation.ts" region="interval" header="Create an observable from a counter"></code-example>

<code-example path="rx-library/src/simple-creation.ts" region="event" header="Create an observable from an event"></code-example>

<code-example path="rx-library/src/simple-creation.ts" region="ajax" header="Create an observable that creates an AJAX request"></code-example>

### Operators

> Operators are functions that build on the observables foundation to enable sophisticated manipulation of collections. For example, RxJS defines operators such as `map()`, `filter()`, `concat()`, and `flatMap()`.

Reviewer 3: ```suggestion
Operators are functions that allow you to manipulate the values observables return. For example, RxJS defines operators such as `map()`, `filter()`, `concat()`, and `flatMap()`.
```

> Operators take configuration options, and they return a function that takes a source observable. When executing this returned function, the operator observes the source observable’s emitted values, transforms them, and returns a new observable of those transformed values. Here is a basic example:

Reviewer 3: ```suggestion
Operators take configuration options, and they return a function that takes a source observable. When executing this returned function, the operator observes the source observable’s emitted values, transforms them, and returns a new observable of those transformed values. For example:
```

<code-example path="rx-library/src/operators.ts" header="Map operator"></code-example>

You can use _pipes_ to link operators together. Pipes let you combine multiple functions into a single function. The `pipe()` function takes as its arguments the functions you want to combine, and returns a new function that, when executed, runs the composed functions in sequence.

<div class="alert is-helpful">

>   Note that, for Angular apps, combining operators with pipes is preferred to _chaining_, which is used in many RxJS examples.

Reviewer 3: ```suggestion
 For Angular applications, it is recommended that you combine operators with pipes.
```

Author: I think we should keep the note about pipes vs chaining, since we refer to RxJS doc for operators, and their examples use chaining.

</div>

> A set of operators applied to an observable is a recipe&mdash;that is, a set of instructions for producing the values you’re interested in. By itself, the recipe doesn’t do anything. You need to call `subscribe()` to produce a result through the recipe.

Reviewer 3: Is recipe in our glossary?

Author: No, it's a general descriptive term that Jason came up with. I'll try to rephrase.

Here’s an example:

<code-example path="rx-library/src/operators.1.ts" header="Standalone pipe function"></code-example>

> The `pipe()` function is also a method on the RxJS `Observable`, so you can use this shorter form to define the same operation:

Reviewer 3: ```suggestion
The `pipe()` function is also a method on the RxJS `Observable`, which means you can use this shorter form to define the same operation:
```

<code-example path="rx-library/src/operators.2.ts" header="Observable.pipe function"></code-example>

### Common operators

> RxJS provides many operators, but only a handful are used frequently. For a full list of operators and usage samples, visit the [RxJS API documentation](https://rxjs.dev/api).
> 
> 
> | Area | Operators |
> | :------------| :----------|
> | Creation |  `from`,`fromEvent`, `of` |
> | Combination | `combineLatest`, `concat`, `merge`, `startWith` , `withLatestFrom`, `zip` |
> | Filtering | `debounceTime`, `distinctUntilChanged`, `filter`, `take`, `takeUntil` |
> | Transformation | `bufferTime`, `concatMap`, `map`, `mergeMap`, `scan`, `switchMap` |
> | Utility | `tap` |
> | Multicasting | `share` |
> 

Reviewer 3: ```suggestion
RxJS provides many operators. The most common operators used in Angular applications are as follows: 

| Area | Operators |
| :------------| :----------|
| Creation |  `from`,`fromEvent`, `of` |
| Combination | `combineLatest`, `concat`, `merge`, `startWith` , `withLatestFrom`, `zip` |
| Filtering | `debounceTime`, `distinctUntilChanged`, `filter`, `take`, `takeUntil` |
| Transformation | `bufferTime`, `concatMap`, `map`, `mergeMap`, `scan`, `switchMap` |
| Utility | `tap` |
| Multicasting | `share` |

For a full list of operators and usage samples, visit the [RxJS API documentation](https://rxjs.dev/api).
```

## Naming conventions for observables

> Because Angular applications are mostly written in TypeScript, you will typically know when a variable is an observable. Although the Angular framework does not enforce a naming convention for observables, you will often see observables named with a trailing “$” sign.

Reviewer 3: ```suggestion
Angular applications are mostly written in TypeScript, which makes it easier to know when a variable is an observable. However, a common naming convention for observables is to add a `$` to the variable name. For example, `response$`. Remember that the Angular framework does not enforce this convention.
```

> This can be useful when scanning through code and looking for observable values. Also, if you want a property to store the most recent value from an observable, it can be convenient to simply use the same name with or without the “$”.

Reviewer 3: ```suggestion
Using this naming convention can help you when you are scanning through code and looking for observable values. 
```
I recommend deleting the last sentence. First we tell people about a convention, then we tell them it's not enforced, then we tell them they may or may not use it? I think it's fine to explain that it's a useful, if unenforced, convention, but that's it.

Author: Rephrased to describe usage in example.

});
</code-example>

In addition to the `error()` handler that you provide on subscription, the RxJS library provides the [`catchError` operator](https://rxjs.dev/api/operators/catchError "RxJS API reference") that lets you handle known errors in the observable code.

> For instance, suppose you have an observable that makes an API request and maps to the response from the server. If the server returns an error or the value doesn’t exist, an error is produced. If you catch this error and supply a default value, your stream continues to process values rather than erroring out.

Reviewer: ```suggestion
For instance, suppose you have an observable that makes an API request and maps to the response from the server. If the server returns an error or the value doesn’t exist, an error is produced. If you catch this error and supply a default value, your stream continues to process values rather than propagating the error.
```

The following example demonstrates the `catchError` operator:

<code-example path="rx-library/src/error-handling.ts" header="catchError operator"></code-example>

### Retrying a failed observable

> Where the `catchError` operator provides a simple path of recovery, the [`retry` operator](https://rxjs.dev/api/operators/retry "RxJS API reference") lets you retry a failed request.

Reviewer: ```suggestion
Whereas the `catchError` operator provides a simple path of recovery, the [`retry` operator](https://rxjs.dev/api/operators/retry "RxJS API reference") lets you retry a failed request.
```

> Use the `retry` operator before the `catchError` operator. It resubscribes to the original source observable, which can then re-run the full sequence of actions that resulted in the error. If this includes an HTTP request, it will retry that HTTP request.

Reviewer: It might be worth mentioning that retrying works by re-subscribing to the original observable, which e.g. for HTTP requests, will result in a new request being fired.

That's also why it's important to `retry()` before `catchError()` or any multicasting operations like `share()`.

<code-example path="rx-library/src/simple-creation.ts" region="promise" header="Create an observable from a promise"></code-example>

<code-example path="rx-library/src/simple-creation.ts" region="interval" header="Create an observable from a counter"></code-example>

<code-example path="rx-library/src/simple-creation.ts" region="event" header="Create an observable from an event"></code-example>

<code-example path="rx-library/src/simple-creation.ts" region="ajax" header="Create an observable that creates an AJAX request"></code-example>

### Operators

Operators are functions that allow you to manipulate the values that observables return.
For example, RxJS defines operators such as `map()`, `filter()`, `concat()`, and `mergeMap()`.

Operators take configuration options, and they return a function that takes a source observable.
When executing this returned function, the operator observes the source observable’s emitted values, transforms them, and returns a new observable of those transformed values.
For example:

<code-example path="rx-library/src/operators.ts" header="Map operator"></code-example>

> You can use _pipes_ to link operators together. Pipes let you combine multiple functions into a single function.

Reviewer: This is a little confusingly worded. There is no named concept of a "pipe" in Rx. `pipe()` is the name of the method we use to transform an observable via one or more operators.

The `pipe()` function takes as its arguments the functions you want to combine, and returns a new function that, when executed, runs the composed functions in sequence.

<div class="alert is-helpful">

>  For Angular applications, it is recommended that you combine operators with pipes (rather than _chaining_, which is used in many RxJS examples).

Reviewer: ```suggestion
 For Angular applications, it is recommended that you combine operators with `pipe()` (rather than _chaining_, which was used in older versions of RxJS).
```

</div>

A set of operators applied to an observable is a set of instructions for producing the values you’re interested in.
The operators require input from the execution of an observable.
You need to call `subscribe()` to execute the observable and produce a result from the operators.

Here’s an example:

> <code-example path="rx-library/src/operators.1.ts" header="Standalone pipe function"></code-example>

Reviewer: Nobody uses the standalone `pipe` function - we should either switch these examples, or just not teach the standalone way at all.

The `pipe()` function is also a method on the RxJS `Observable`, which means you can use this shorter form to define the same operation:

<code-example path="rx-library/src/operators.2.ts" header="Observable.pipe function"></code-example>

### Common operators

RxJS provides many operators. The following are commonly used in Angular applications:

| Area | Operators |
| :------------| :----------|
| Creation |  `from`,`fromEvent`, `of` |
| Combination | `combineLatest`, `concat`, `merge`, `startWith` , `withLatestFrom`, `zip` |
| Filtering | `debounceTime`, `distinctUntilChanged`, `filter`, `take`, `takeUntil` |
| Transformation | `bufferTime`, `concatMap`, `map`, `mergeMap`, `scan`, `switchMap` |

> | Utility | `tap` |

Reviewer: It might actually be useful to have a section on debugging RxJS code, where `tap` can come in handy for logging or adding breakpoints during an RxJS sequence.

Author: Let me know if you can provide source for such a section.

| Multicasting | `share` |

For a full list of operators and usage samples, visit the [RxJS API documentation](https://rxjs.dev/api).

## Naming conventions for observables

Angular applications are mostly written in TypeScript, which makes it easier to know when a variable is an observable.
However, a common naming convention for observables is to add a `$` to the variable name; for example, `response$`.
Remember that the Angular framework does not enforce this convention.

> Using this naming convention can help you when you are scanning through code and looking for observable values, or when you want to associate an observable with a property that stores its returned values.

Reviewer: I've also found it useful when you have both an Observable and its value in the same namespace. For example, within a template, the following might be common:
```html
<div *ngFor="let user of users$ | async as users">...</.div>
```
where `users$` is the Observable that's passed to the `async` pipe, and the unwrapped/emitted value is aliased as `users`. This saves you from having to get creative with naming for both the sync and async versions of a value.

You can import these from the `rxjs` library in your Angular application, and use them to accomplish the following kinds of tasks:

* Convert existing code for async operations into observables
* Iterate through the values in a stream
* Map values to different types
* Filter streams
* Compose multiple streams

### Observable creation functions

RxJS offers a number of functions that you can use to create new observables. These functions help you create observables from things such as events, timers, promises, and so on. For example:

<code-example path="rx-library/src/simple-creation.ts" region="promise" header="Create an observable from a promise"></code-example>

<code-example path="rx-library/src/simple-creation.ts" region="interval" header="Create an observable from a counter"></code-example>

<code-example path="rx-library/src/simple-creation.ts" region="event" header="Create an observable from an event"></code-example>

<code-example path="rx-library/src/simple-creation.ts" region="ajax" header="Create an observable that creates an AJAX request"></code-example>

### Operators

Operators are functions that allow you to manipulate the values that observables return.

> For example, RxJS defines operators such as `map()`, `filter()`, `concat()`, and `flatMap()`.

Reviewer 2: flatMap is deprecated, please use mergeMap
